### PR TITLE
SDK Spec: Backend-only

### PR DESCRIPTION
**Pull request by LekoArts**
> [!IMPORTANT]
> 🔎 Previews:
>
> - https://clerk.com/docs/pr/1394/references/sdk/backend-only

This PR will be merged into https://github.com/clerk/clerk-docs/pull/1344 and operates in the same context of "SDK Spec". Companion to https://github.com/clerk/clerk-docs/pull/1368

Fixes ECO-110
Fixes ECO-100

### CODE REVIEW

File: docs/references/sdk/backend-only.mdx [hunk 1]
```diff
@@ -1,7 +1,192 @@
 ---
-description: TODO
+description: A reference for implementing a backend-only Clerk SDK
 ---
 
 # Backend-only SDK
 
-TODO
+Generally speaking, you can either create a backend SDK that implements all BAPI endpoints or an SDK for an existing backend framework. You'll find the implementation steps below divided into these two categories.
```
Comment (alexisintech) [inline docs/references/sdk/backend-only.mdx:7]:
```suggestion
When creating a backend-only SDK, you have two options for implementing the BAPI endpoints: either [develop a backend SDK that encompasses all BAPI endpoints](#implementation-bapi) or [create an SDK tailored for an existing backend framework.](#implementation-node-js-backend-framework)
```

File: docs/references/sdk/backend-only.mdx [hunk 2]
```diff
@@ -1,7 +1,192 @@
...
+
+`@clerk/backend` is built for Node.js/V8 isolates (Cloudflare Workers, Vercel Edge Runtime, etc.). It’s the foundational package for all JavaScript Backend SDKs and works across all JavaScript runtimes. By using `@clerk/backend` you can be sure to communicate with Clerk’s BAPI in a correct and secure way.
+
+> [!NOTE]
+> The code blocks below will be written in pseudo-code. If you're looking for real-world examples, have a look at [`@clerk/fastify`](https://github.com/clerk/javascript/tree/main/packages/fastify) and [`@clerk/express`](https://github.com/clerk/javascript/tree/main/packages/express).
+
+<Steps>
+  ### Create a Clerk client
+
+  Use `createClerkClient` from `@clerk/backend` to create your default Clerk client which will be used for the middleware.
```
Comment (alexisintech) [inline docs/references/sdk/backend-only.mdx:47]:
```suggestion
  Use [`createClerkClient`](/docs/references/backend/overview#usage) from `@clerk/backend` to create your default Clerk client which will be used for the middleware.
```

File: docs/references/sdk/backend-only.mdx [hunk 3]
```diff
@@ -1,7 +1,192 @@
...
+    apiVersion: API_VERSION,
+    jwtKey: JWT_KEY,
+    userAgent: `${PACKAGE_NAME}@${PACKAGE_VERSION}`,
+    sdkMetadata: SDK_METADATA,
+  })
+  ```
+
+  ### Create your middleware/plugin
+
+  Inside the middleware, you’ll use the user-provided Clerk client (or use the default one created in the previous step) and authenticate the request. `authenticateRequest` returns the request state which the middleware should set into its context (or similar). This way other helpers can access it later in the chain.
```
Comment (alexisintech) [inline docs/references/sdk/backend-only.mdx:75]:
```suggestion
  Inside the middleware, you’ll use the user-provided Clerk client (or use the default one created in the previous step) and authenticate the request. [`authenticateRequest`](/docs/references/backend/authenticate-request) returns the request state, which the middleware should set into its context (or similar). This way other helpers can access it later in the chain.
```

File: docs/references/sdk/backend-only.mdx [hunk 4]
```diff
@@ -1,7 +1,192 @@
...
+
+    return context.json({ message: 'Logged in', userId: auth.userId })
+  })
+  ```
+
+  ### Create a `requireAuth` helper
+
+  This utility will require auth requests for user authenticated or authorized requests. An HTTP `401` status code is returned for unauthorized requests.
+
+  It’ll depend on your framework if that utility needs to take the hasPermission as an argument or can be chained as a handler after it.
```
Comment (alexisintech) [inline docs/references/sdk/backend-only.mdx:123]:
```suggestion
  It’ll depend on your framework if that utility needs to take `hasPermission` as an argument or can be chained as a handler after it.
```

Comment (alexisintech) [reply docs/references/sdk/backend-only.mdx:123]: can I have more info on `hasPermission` ?

File: docs/references/sdk/backend-only.mdx [hunk 5]
```diff
@@ -1,7 +1,192 @@
...
+  ```ts {{ filename: 'index.ts', mark: [3, 16] }}
+  const app = new Framework()
+
+  app.get('/path', requireAuth())
+
+  // Or with checking the permissions
+
+  const hasPermission = (context, next) => {
+    const auth = getAuth(context)
+    if (!auth.has({ permission: 'permission' })) {
```
Comment (alexisintech) [inline docs/references/sdk/backend-only.mdx:149]: is that what you meant by `hasPermission` ? the `has({ permission })` ?

Comment (LekoArts) [reply docs/references/sdk/backend-only.mdx:154]: Thanks for asking, I made some changes and hopefully clarified some things!

Essentially I described https://clerk.com/docs/references/nextjs/auth-object and its `has()`. It's not unique to Next.js (so maybe in the wrong place?) and for now I linked to it.

Let me know if the updated text makes more sense

File: docs/references/sdk/backend-only.mdx [hunk 6]
```diff
@@ -1,7 +1,197 @@
...
+    }
+    return next()
+  }
+
+  app.get('/path', requireAuth(), hasPermission())
+  ```
+
+  ### Add handshake support
+
+  Add checks for the `headers` to apply them from the request to the response and handle and redirect headers.
```
Comment (alexisintech) [inline docs/references/sdk/backend-only.mdx:166]: i think there's a typo in this sentence towards the end ?

File: docs/references/sdk/backend-only.mdx [hunk 7]
```diff
@@ -1,7 +1,192 @@
...
+Generally speaking, you can either create a backend SDK that implements all BAPI endpoints or an SDK for an existing backend framework. You'll find the implementation steps below divided into these two categories.
+
+The source of truth for all BAPI endpoints is the [BAPI OpenAPI spec](https://clerk.com/docs/reference/backend-api). For Node.js backend frameworks, use the [JavaScript Backend SDK](/docs/references/backend/overview).
+
+## Expected features
+
+- User only needs to provide their [secret key](/docs/references/sdk/terminology)
+- Centralized request authentication (e.g. in a middleware or plugin)
+- Give access to the instance of Backend API client (so that users can use all methods)
+- User should be able to do auth for roles and permissions
```
Comment (jescalan) [inline docs/references/sdk/backend-only.mdx:16]: What does this mean

File: docs/references/sdk/backend-only.mdx [hunk 8]
```diff
@@ -1,7 +1,192 @@
...
+> [!IMPORTANT]
+> BAPI has [rate limits](/docs/backend-requests/resources/rate-limits) to help protect users against brute-force attacks or stop abuse of Clerk's platform. Be sure to include a backoff mechanism into your fetching logic and respect the `Retry-After` header to gracefully handle any active rate limits.
+
+If you're using `@clerk/backend` to build an SDK for an existing framework, these additional features are expected:
+
+- User should be able to use all `@clerk/backend` options
+
+## Optional features
+
+- User should be able to enforce auth on individual routes
```
Comment (jescalan) [inline docs/references/sdk/backend-only.mdx:27]: Would be helpful to drop an example in here like the express sdk perhaps

File: docs/references/sdk/backend-only.mdx [hunk 9]
```diff
@@ -1,7 +1,192 @@
...
+- Use singleton pattern to only create one instance of Clerk backend client
```
Comment (jescalan) [inline docs/references/sdk/backend-only.mdx:28]:
```suggestion
- Use singleton pattern to only create a pre-configured instance of Clerk backend client
```

File: docs/references/sdk/backend-only.mdx [hunk 10]
```diff
@@ -1,7 +1,192 @@
...
+    }
+    return next()
+  }
+
+  app.get('/path', requireAuth(), hasPermission())
+  ```
+
+  ### Add handshake support
+
+  Add checks for the `headers` to apply them from the request to the response and handle and redirect headers.
```
Comment (jescalan) [inline docs/references/sdk/backend-only.mdx:166]: This is hard to grapple with if you don't have any understanding of what handshake is, or why you would need to do something like this. We could file it as a todo, as I'm not sure that writing an explanation of handshake for the docs is in scope for this PR, but this section is effectively just confusing for anyone who would realistically be trying to do anything with it.

Comment (LekoArts) [reply docs/references/sdk/backend-only.mdx]: Yeah, correct 👍 

I also noticed this knowledge gap but decided against working on it for this PR. In the internal Notion doc I mentioned that we probably should bring https://github.com/clerk/clerk-docs/pull/577 over the finish line and add any new info we have on handshake (that was written internally in the meantime).

**Pull request merged**